Fix the Oomd::updateContext() exception's in Oomd.cpp, Fs.cpp

Hyperion101010 opened the pull request:
i added exception handling in the part where there was a possibility of abort due to error ,please review

danobi commented:
Couple things:
* Can you remove all the other code that isn't related to the updateContext bug?
* Can you read through https://chris.beams.io/posts/git-commit/ and make your commit messages more descriptive?
* Are you sure this code compiles?

Hyperion101010 commented:
i am doubtful about it 
i have changed it ,added exceptions and it builds  and checks fine in "ninja test"
just please check it first hand 
it took me time becuase i am new to using git 
i am new to everything 
i am preparing for gsoc for mozilla , i hope you don't mind :)

danobi commented on install/build.ninja:
```diff
@@ -0,0 +1,182 @@
+# This is the build file for project "oomd"
```
This file is a build artifact. It should not be checked in.

danobi commented on Oomd.h:
```diff
@@ -78,6 +78,10 @@ class Oomd {
   std::vector<std::unique_ptr<Cgroup>> cgroups_;
 
   bool warned_io_pressure_{false};
+  /*bool warned_memcurrent{false};
```
Commiting commented out code is generally a bad idea. Such code is not compiled or tested and could be invalid by the time someone tries to uncomment.

danobi commented on Oomd.cpp:
```diff
@@ -162,20 +189,23 @@ int Oomd::run() {
     }
 
-    for (auto& cgroup : cgroups_) {
+    for (auto &cgroup : cgroups_)
```
Try not to make styling changes -- it adds too much noise to the review

danobi commented on util/Fs.cpp:
```diff
@@ -184,7 +186,10 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
-  OCHECK(lines.size() == 1);
+  if (!(lines.size() == 1))
+  {
+    throw "readMemcurrent (short lived process)";
```
Throwing `char *` is usually not a good idea. Try creating a generic exception type in `include/Types.h` and throw that instead. That way, different errors can be handled or propagated up the stack.

danobi commented on install/compile_commands.json:
```diff
@@ -0,0 +1,62 @@
+[
```
Don't check this in either.

Hyperion101010 commented:
danobi i have added the required changes 
you can review this patch .

danobi commented:
It looks like you merged your local branch into this PR. Can you squash all the commits and remove the noise I mentioned earlier?

http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Hyperion101010 commented:
I’ll do it in a day or two 
I am busy for a while 
Besides is it the code that you were expecting

danobi commented:
There's a few style suggestions and best practice advice, but yeah, it's looking much better.

danobi commented:
Can you make sure the build passes?

Hyperion101010 commented:
i am working on it , it'll be done in a day or two .

Hyperion101010 commented:
the thing happened was as you changed the code , my first patch lost the changes in other files 
thus i had to rewrite it in a more than one file .

danobi commented on include/Types.h:
```diff
@@ -18,9 +18,19 @@
 #include <cstdint>
+#include <string.h>
+//#include <exception.h>
+#include <bits/stdc++.h>
```
It's usually not a good idea to include all the stl libraries in one go. Prefer to include as little as possible.

danobi commented on include/Types.h:
```diff
@@ -18,9 +18,19 @@
 
 #include <cstdint>
+#include <string.h>
+//#include <exception.h>
```
`#include <exception>` doesn't work?

danobi commented on include/Types.h:
```diff
@@ -18,9 +18,19 @@
 
 namespace Oomd {
 
+class short_lived_exception : public std::exception {
```
On second thought, I don't think we need a custom exception type if we're only passing string messages. It might be useful if we needed more context (eg from a vector), but we don't. It's probably cleaner to throw an `std::runtime_error`.

danobi commented on util/Fs.cpp:
```diff
@@ -252,7 +252,10 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
-    OCHECK(lines.size() == 3);
+    if (lines.size() != 3) {
+      short_lived_exception exception_var( "(short lived process ) read Mempressure");
+      throw &exception_var;
```
Throwing a pointer to something on the stack is probably not a good idea. Although it might still work, this is undefined behavior. Prefer doing this instead:
```
throw std::runtime_error("readMempressure could not read filesystem");
```

danobi commented on util/Fs.cpp:
```diff
@@ -266,7 +269,10 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
-  OCHECK(lines.size() == 1);
+  if (lines.size()!=1){
+    short_lived_exception exception_var("(short lived process ) read Memcurrent");
+    throw &exception_var;
```
Same as above

danobi commented on util/Fs.cpp:
```diff
@@ -289,7 +295,10 @@ ResourcePressure Fs::readMempressure(const std::string& path) {
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    short_lived_exception exception_var("(short lived process ) read Memlow");
+    throw &exception_var;
```
Same as above

danobi commented on util/Fs.cpp:
```diff
@@ -298,7 +307,10 @@ int64_t Fs::readMemlow(const std::string& path) {
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    short_lived_exception exception_var("(short lived process ) read SwapCurrent");
+    throw &exception_var;
```
Same as above

danobi commented:
I noticed you also tried to fix some compiler warnings. For stuff like that, it's best done in another patch. It's best to have each commit do only one thing.

Hyperion101010 commented:
good to go now ! that was a lot of work/time for my first patch for this library .
But it is understandable cause it took me not more than 1hr to figure out and write code 
And 23 days to fit in the norms , how to use git and all !

danobi commented on config/JsonConfigParser.cpp:
```diff
@@ -74,7 +74,7 @@ Oomd::Config2::IR::DetectorGroup parseDetectorGroup(
   }
 
-  for (int i = 0; i < detector_group.size(); ++i) {
+  for (int i = 0; i < (signed)detector_group.size(); ++i) {
```
I fixed the build warnings on master. Can you rebase onto those changes?

danobi commented on Oomd.cpp:
```diff
@@ -81,14 +81,17 @@ bool Oomd::updateContextCgroup(
 
     // Can't extract much info if memory controller isn't enabled
     return false;
+  };
```
```suggestion
  };
  
```

danobi replied:
Could you please add a new line here like the suggestion change says?

danobi commented on Oomd.cpp:
```diff
@@ -81,14 +81,17 @@ bool Oomd::updateContextCgroup(
+    swap_current = Fs::readSwapCurrent(absolute_cgroup_path);
+  } catch (const std::exception& ex) {
+    OLOG << ex.what();
   }
```
```suggestion
  }
  
```

danobi commented on util/Fs.cpp:
```diff
@@ -252,7 +252,9 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
     // full 0.00 0.03 0.05
-    OCHECK(lines.size() == 3);
+    if (lines.size() != 3) {
+      throw std::runtime_error("could not read Mempressure process terminated");
```
I think this would be more informative, as there are usually many processes in a cgroup:
```suggestion
      throw std::runtime_error("cgroup is no longer valid: " + path);
```

danobi commented on util/Fs.cpp:
```diff
@@ -266,7 +268,9 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
   auto lines = readFileByLine(path + "/" + kMemCurrentFile);
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    throw std::runtime_error("could not read Memcurrent process terminated");
```
Same as above

danobi commented on util/Fs.cpp:
```diff
@@ -289,7 +293,9 @@ ResourcePressure Fs::readMempressure(const std::string& path) {
   auto lines = readFileByLine(path + "/" + kMemLowFile);
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    throw std::runtime_error("could not read Memlow process terminated");
```
Same as above

danobi commented on util/Fs.cpp:
```diff
@@ -298,7 +304,9 @@ int64_t Fs::readMemlow(const std::string& path) {
   auto lines = readFileByLine(path + "/" + kMemSwapCurrentFile);
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    throw std::runtime_error("could not read SwapCurrent process terminated");
```
Same as above

Hyperion101010 commented:
done ! 
i think now its good 
But anyways i was thinking about optimizing the method you use for oomd identifying the processes .

Hyperion101010 commented:
now review !

Hyperion101010 commented on util/Fs.cpp:
```diff
@@ -252,7 +252,9 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
     // full 0.00 0.03 0.05
-    OCHECK(lines.size() == 3);
+    if (lines.size() != 3) {
+      throw std::runtime_error("cgroups are not valid:" + path);
```
here i can understand , there can be multiple paths

Hyperion101010 commented on util/Fs.cpp:
```diff
@@ -266,7 +268,9 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
   auto lines = readFileByLine(path + "/" + kMemCurrentFile);
-  OCHECK(lines.size() == 1);
+  if (lines.size() != 1) {
+    throw std::runtime_error("cgroups are not valid" + path);
```
why here and below from here ,
here there can only be one or no path 
so

danobi replied:
I was suggesting a grammatical fix for the exception message. I'm not sure what you mean by "one or no path".

danobi commented on util/Fs.cpp:
```diff
@@ -252,7 +252,14 @@ ResourcePressure Fs::readRespressure(const std::string& path) {
-    OCHECK(lines.size() == 3);
+    if (lines.size() != 3) {
+      std::string exception_error;
+      for(unsigned int i=0;i<lines.size();i++)
```
I don't think the loop is necessary. It was fine the way it was before. The change I suggested was exactly the change I wanted you to make. No more, no less.

Hyperion101010 commented:
is there a need to update tests for the new mem_stat feature you added?

danobi commented:
I think you checked in some patch files by accident.
